### ISSUE TITLE
New tutorial on the most fundamental functionalities 

### ISSUE THREAD

**Issue by vegapit** [LaurentMazare/tch-rs#75]
Chose promise, chose due...

Here is a [tutorial ](http://vegapit.com/article/how-to-use-torch-in-rust-with-tch-rs) I wrote covering the most fundamental use cases for this library.

Let me know if I got anything wrong, cheers.

**Comment by vegapit** [comment]
I shared it on social media and it is getting a fair amount of traction. Your user base is growing fast.

**Comment by LaurentMazare** [comment]
I've added a reference to this tutorial in the main readme, happy to see more material there if you feel like writing on other functionalities.

**Comment by vegapit** [comment]
Sure, I am happy to help you further with that. I am still in the process of translating more PyTorch code over to Rust so I am, and will be, looking into the inner workings of the library further. Let me know if there is a particular aspect that you see as important to dig into.

From what I can gather, the library is appealing to both Pytorch and Rust users, so I will shape the tutorials to help people falling into either categories. As you have probably guessed, my field of expertise is quantitative trading/finance so my examples will most likely revolve around this subject if it is ok with you.

**Comment by vegapit** [comment]
Quick question: If I have defined 2 independent modules that I would like to backpropagate in succession,  can I define them on the same `VarStore` and the same `Optimizer` or each one has to have their own?

**Comment by LaurentMazare** [comment]
You can certainly use the same var-store/optimizer. You can use different paths if you want the variables namespaces to be a bit isolated but that's not even mandatory. E.g.

```rust
let vs = nn::VarStore::new(Device::Cpu);
let module1 = module1(vs.root() / "mod1", ...);
let module2 = module1(vs.root() / "mod2", ...);
// This optimizes over all trainable variables for both modules.
let opt = nn::Sgd::default().build(&vs, 1e-3).unwrap();
```

**Comment by vegapit** [comment]
Say I have 2 losses and, given that the two modules use independent branches of the computation graph, would best practice be to have 1 VarStore for both modules but one optimiser each? I am just not sure if the gradients would make sense if I called `backward_step` twice if they were defined on the same optimizer.

**Comment by LaurentMazare** [comment]
For the [gan example](https://github.com/LaurentMazare/tch-rs/blob/master/examples/gan/main.rs) we use two var-store/optimizers but this is because we only want to backpropagate to the generator or discrimator weights and not to both together (note that we use `freeze` and `unfreeze` for this).
If you also want to ensure that the weights from only one module get updated during the backprop step and not the weights from the other module, using two var-stores and optimizers is a good idea.
However if you want to update the weights of both modules on each backprop step then a single optimizer should be fine. The `backward_step` [implementation](https://github.com/LaurentMazare/tch-rs/blob/0625cb9782a6f4ab74c35a889a9501dfa2faec05/src/nn/optimizer.rs#L165-L168) actually starts by zeroing the gradients so the gradients for the second loss should be gone at this point.

**Comment by vegapit** [comment]
Understood, thank you. I am experimenting by running actor/critic resolutions on multiple agents at the same time and the `backward_step` calls were indeed erasing the gradients. I was not sure if the dedicated VarStore & Optimizer approach was a recommended one, but it seems to be working fine without any noticeable pressure on the RAM.

**Comment by vegapit** [comment]
New tutorial on tch using a reinforcement learning example: [link](http://vegapit.com/article/rl-with-torch-in-rust-kuhn-poker). Let me know if you spot any problems with it before I share it with a wider audience.

**Comment by LaurentMazare** [comment]
Looks very nice, thanks!
I haven't dived a lot into it but here are some very minor points (I may be wrong about a couple of them):
- "I have modelled" -> "I have modeled".
- "and organise them" -> "and organised them".
- "only define whether" -> "only defines whether".
- "both consists" -> "both consist".
- When randomly generating the deck, you may be able to simplify things a bit:
```rust
    let shuffle = Tensor::of_slice(&[1.0, 1.0, 1.0]).multinomial(2, false);
    let cards = Vec::<i64>::from(&shuffle);
```
- You seem to be building tensors out of vectors via `Tensor::stack`. maybe you could use `Tensor::of_slice` directly (and then `.view(...)` if you want to tweak the shape).
- Rather than `scatter_(..., &Tensor::from(1f32))` you should be able to use `scatter_1(..., 1.0)`.

**Comment by vegapit** [comment]
Superb, I have corrected all but these two points:

* "modelled" is actually the right spelling.
* Replacing `Tensor::stack` by `Tensor::of_slice` does not seem to work in this case. The problem is that the inputs are `Vec<Tensor>` and the `Tensor::of_slice` method does not support a slice of `Tensor` as input. It is not a problem as the use of `stack` makes the code very clear and concise.

I will pass the article on

**Comment by vegapit** [comment]
For some reason, this new tutorial did not spark much interest at all compared to the previous one. Was it too technical? Or is Reinforcement Learning not as popular as Supervised Learning? I am not quite sure.

**Comment by LaurentMazare** [comment]
I would think that RL is pretty popular and your post was interesting but some part of the success of a twitter post is stochastic especially if you don't have lots of followers: if your post gets retweeted by someone who has lots of followers then there is some kind of snowball effect around it but if you're unlucky then this may just not happen.
After that maybe having more visual content helps (e.g. some video visualization) but I'm not sure how well suited this would be in this case.

**Comment by vegapit** [comment]
Yes, you are right. I have just checked and the difference is that the second one was not retweeted by @Pytorch. Oh well, it is what it is then...
